core/src/Kokkos_Graph.hpp

```diff
@@ -0,0 +1,191 @@
+
+template <class ExecutionSpace>
+struct KOKKOS_ATTRIBUTE_NODISCARD Graph {
+ public:
+  //----------------------------------------------------------------------------
+  // <editor-fold desc="public member types"> {{{2
+
+  using execution_space = ExecutionSpace;
+  using graph           = Graph;
+  using graph_builder   = GraphBuilder<ExecutionSpace>;
+
+  // </editor-fold> end public member types }}}2
+  //----------------------------------------------------------------------------
+
+ private:
+  //----------------------------------------------------------------------------
+  // <editor-fold desc="friends"> {{{2
+
+  friend struct Kokkos::Impl::GraphAccess;
+
+  // </editor-fold> end friends }}}2
+  //----------------------------------------------------------------------------
+
+  //----------------------------------------------------------------------------
+  // <editor-fold desc="private data members"> {{{2
+
+  using impl_t                       = Kokkos::Impl::GraphImpl<ExecutionSpace>;
+  std::shared_ptr<impl_t> m_impl_ptr = nullptr;
+
+  // </editor-fold> end private data members }}}2
+  //----------------------------------------------------------------------------
+
+  //----------------------------------------------------------------------------
+  // <editor-fold desc="private ctors"> {{{2
+
+  // Note: only create_graph() uses this constructor, but we can't just make
+  // that a friend instead of GraphAccess because of the way that friend
+  // function template injection works.
+  explicit Graph(std::shared_ptr<impl_t> arg_impl_ptr)
+      : m_impl_ptr(std::move(arg_impl_ptr)) {}
+
+  // </editor-fold> end private ctors }}}2
+  //----------------------------------------------------------------------------
+
+ public:
+  ExecutionSpace const& get_execution_space() const {
+    return m_impl_ptr->get_execution_space();
+  }
+
+  void submit() const {
+    KOKKOS_EXPECTS(bool(m_impl_ptr))
+    (*m_impl_ptr).submit();
+  }
+};
+
+// </editor-fold> end Graph }}}1
+//==============================================================================
+
+//==============================================================================
+// <editor-fold desc="when_all"> {{{1
+
+template <class... PredecessorRefs>
+// constraints (not intended for subsumption, though...)
+//   ((remove_cvref_t<PredecessorRefs> is a specialization of
+//        GraphNodeRef with get_root().get_graph_impl() as its GraphImpl)
+//      && ...)
+auto when_all(PredecessorRefs&&... arg_pred_refs) {
+  // TODO @graph @desul-integration check the constraints and preconditions
+  //                                once we have folded conjunctions from
+  //                                desul
+  static_assert(sizeof...(PredecessorRefs) > 0,
+                "when_all() needs at least one predecessor.");
+  auto graph_ptr_impl =
+      Kokkos::Impl::GraphAccess::get_graph_weak_ptr(
+          std::get<0>(std::forward_as_tuple(arg_pred_refs...)))
+          .lock();
```

Comment:
I guess at some point we could check in debug whether they all belong to the same graph? Or does that happen implicitly?

Reply:
I.e. does that happen in create_aggregate_ptr?

Reply:
Yeah, at some point we should check this, but it's a bigger change than I wanted to do here (it needs some of the kinds of tools we'll be putting in to desul that we don't have yet), it's hard to accidentally do (since the whole graph is built in a single closure), and it's not worth the trouble in the first version of this feature.

Reply:
agreed.